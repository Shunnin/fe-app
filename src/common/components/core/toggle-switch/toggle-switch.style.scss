@import 'style/foundation/config';

.toggle-switch {
  position: relative;
  display: inline-block;
  width: 50px;
  height: 20px;

  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  transition: 0.4s;
  border-radius: $b-border-radius-round;
  background-color: $c-c7;

  &__on,
  &__off {
    color: $c-wh;
    position: absolute;
    transform: translate(-50%, -50%);
    top: 50%;
    left: 50%;
    font-size: $b-font-size;
    font-weight: $b-font-weight-medium;
    display: flex;
    svg {
      width: 20px;
      fill: $c-wh;
    }
  }

  &__on {
    padding-right: $b-space;
  }

  &__off {
    padding-left: $b-space;
  }

  & &__indicator {
    position: absolute;
    height: 16px;
    width: 18px;
    left: 2px;
    top: 2px;
    background-color: $c-wh;
    transition: 0.4s;
    transform: translateX(0px);
    border-radius: 44px;

    &.is-toggled {
      transform: translateX(28px);
    }
  }
}
