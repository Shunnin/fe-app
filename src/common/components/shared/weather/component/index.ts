export { Temperature } from './temperature/temperature.component';
export { WeatherCard } from './weather-card/weather-card.component';
export { WeatherItem } from './weather-item/weather-item.component';
export * from './weather-icon';
