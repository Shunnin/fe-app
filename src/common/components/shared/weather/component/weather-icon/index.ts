export { WeatherIcon } from './weather-icon.component';
export * from './weather-icon.constant';
