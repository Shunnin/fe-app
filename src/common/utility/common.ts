export const getSevenDaysInWeek = (today: number = Date.now()): string[] => {
  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  const sevenDays = [];

  // eslint-disable-next-line no-magic-numbers
  for (let i = 1; i <= 7; i++) {
    // eslint-disable-next-line no-magic-numbers
    sevenDays.push(days[new Date(today + i * 24 * 60 * 60 * 1000).getDay()]);
  }

  return sevenDays;
};
