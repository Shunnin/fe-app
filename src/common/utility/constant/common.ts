import { noop, transform } from 'lodash-es';

export const TEMP_UNIT = Object.freeze({
  CELSIUS: 'CELSIUS',
  FAHRENHEIT: 'FAHRENHEIT',
});

export const TEMP_UNIT_VALUE = Object.freeze({
  [TEMP_UNIT.CELSIUS]: true,
  [TEMP_UNIT.FAHRENHEIT]: false,
});

export const TEMP_UNIT_TOGGLE_VALUE = transform(TEMP_UNIT_VALUE, (result, value, key) => (result[value] = key));

export const TEMP_UNIT_MAPPING_LABEL = Object.freeze({
  true: 'C',
  false: 'F',
});

export const EMPTY_FUNC = noop;
export type IFunction = ReturnType<typeof EMPTY_FUNC>;

export const DEFAULT_LOCALE = 'en';

export const WEATHER_API_KEY = process.env.REACT_APP_WEATHER_API_KEY || '20571ab45c74dc2a1897b60c5b8047a1';
