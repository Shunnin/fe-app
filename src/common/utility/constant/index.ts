export * from './number';
export * from './common';
export * from './url';
