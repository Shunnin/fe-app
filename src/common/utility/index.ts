export * from './route';
export * from './constant';
export * from './promise';
export * from './temperature';
export * from './common';
export * from './url';
