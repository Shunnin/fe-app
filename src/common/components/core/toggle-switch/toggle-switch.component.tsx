import { FC, useState, useCallback, useMemo } from 'react';
import classNames from 'classnames';
import { get } from 'lodash-es';

import { EMPTY_FUNC, IFunction } from '../../../utility';
import './toggle-switch.style';

interface IMappingLabel {
  true: string;
  false: string;
}

interface IToggleSwitchProps {
  name?: string;
  defaultValue?: boolean;
  mappingLabel?: IMappingLabel;
  onClick: IFunction;
}

export const ToggleSwitch: FC<IToggleSwitchProps> = ({
  name,
  defaultValue = false,
  mappingLabel,
  onClick = EMPTY_FUNC,
}) => {
  const [toggled, setToggled] = useState(defaultValue);

  const handleClickSwitch = useCallback(() => {
    setToggled(checked => !checked);
    onClick(!toggled);
  }, [onClick, setToggled, toggled]);

  const labelToggle = useMemo(() => get(mappingLabel, toggled, ''), [mappingLabel, toggled]);

  return (
    <div className="toggle-switch clickable" role="button" tabIndex={0} aria-label={name} onClick={handleClickSwitch}>
      {toggled && <span className="toggle-switch__on">{labelToggle}</span>}
      {!toggled && <span className="toggle-switch__off">{labelToggle}</span>}
      <div className={classNames('toggle-switch__indicator', { 'is-toggled': toggled })} />
    </div>
  );
};
