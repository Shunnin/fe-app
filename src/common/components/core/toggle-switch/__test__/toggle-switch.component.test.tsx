import { render, screen, fireEvent } from '@testing-library/react';

import { EMPTY_FUNC } from '../../../../utility/constant';

import { ToggleSwitch } from '../toggle-switch.component';

describe('<ToggleSwitch />', () => {
  const props = {
    name: 'Test toggle switch',
    defaultValue: false,
    onClick: EMPTY_FUNC,
  };

  it('renders and matches the snapshot with defaultValue is false', () => {
    const { container } = render(<ToggleSwitch {...props} />);

    const toggleSwitchEl = container.firstChild;
    const toggleAriaLabelName = screen.getByLabelText(props.name, { selector: 'div' });

    expect(toggleSwitchEl).toMatchSnapshot();
    expect(toggleAriaLabelName).toBeInTheDocument();
    expect(toggleSwitchEl.firstChild).toHaveClass('toggle-switch__off');
    expect(toggleSwitchEl.lastChild).not.toHaveClass('is-toggled');
  });

  it('renders and matches the snapshot with defaultValue is true', () => {
    const { container } = render(<ToggleSwitch {...props} defaultValue />);

    const toggleSwitchEl = container.firstChild;
    const toggleAriaLabelName = screen.getByLabelText(props.name, { selector: 'div' });

    expect(toggleSwitchEl).toMatchSnapshot();
    expect(toggleAriaLabelName).toBeInTheDocument();
    expect(toggleSwitchEl.firstChild).toHaveClass('toggle-switch__on');
    expect(toggleSwitchEl.lastChild).toHaveClass('is-toggled');
  });

  it('handles clicks', () => {
    const onClickMock = jest.fn();
    const { container } = render(<ToggleSwitch {...props} onClick={onClickMock} />);
    const toggleSwitchEl = container.firstChild;

    fireEvent.click(toggleSwitchEl);

    expect(onClickMock).toHaveBeenCalledTimes(1);
    expect(toggleSwitchEl.firstChild).toHaveClass('toggle-switch__on');
    expect(toggleSwitchEl.lastChild).toHaveClass('is-toggled');

    fireEvent.click(toggleSwitchEl);

    expect(onClickMock).toHaveBeenCalledTimes(2);
    expect(toggleSwitchEl.firstChild).toHaveClass('toggle-switch__off');
    expect(toggleSwitchEl.lastChild).not.toHaveClass('is-toggled');
  });
});
