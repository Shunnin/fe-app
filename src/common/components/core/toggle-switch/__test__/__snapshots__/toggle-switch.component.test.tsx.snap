// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleSwitch /> renders and matches the snapshot with defaultValue is false 1`] = `
<div
  aria-label="Test toggle switch"
  class="toggle-switch clickable"
  role="button"
  tabindex="0"
>
  <span
    class="toggle-switch__off"
  />
  <div
    class="toggle-switch__indicator"
  />
</div>
`;

exports[`<ToggleSwitch /> renders and matches the snapshot with defaultValue is true 1`] = `
<div
  aria-label="Test toggle switch"
  class="toggle-switch clickable"
  role="button"
  tabindex="0"
>
  <span
    class="toggle-switch__on"
  />
  <div
    class="toggle-switch__indicator is-toggled"
  />
</div>
`;
